src/library/scala/collection/Iterator.scala

```diff
@@ -1189,6 +1223,80 @@ object Iterator extends IterableFactory[Iterator] {
       } else Iterator.empty.next()
     }
   }
+
+  private final class TakeRightIterator[A](private[this] var underlying: Iterator[A], buflen: Int) extends AbstractIterator[A] {
+    private[this] var len: Int = -1
+    private[this] var pos: Int = 0
+    private[this] var buf: Array[AnyRef] = _
+    def init(): Unit = if(buf eq null) {
+      buf = new Array[AnyRef](buflen)
+      len = 0
+      while(underlying.hasNext) {
+        buf(pos) = underlying.next().asInstanceOf[AnyRef]
+        pos = (pos + 1) % buflen
+        len += 1
+      }
+      underlying = null
+      if(len > buflen) len = buflen
+      pos = (pos - len) % buflen
+      if(pos < 0) pos += buflen
+    }
+    override def knownSize = len
+    def hasNext: Boolean = {
+      init()
+      len > 0
+    }
+    def next(): A = {
+      init()
+      if(len == 0) Iterator.empty.next()
+      else {
+        val x = buf(pos).asInstanceOf[A]
+        pos = (pos + 1) % buflen
+        len -= 1
+        x
+      }
+    }
+    override def drop(n: Int): Iterator[A] = {
+      init()
+      if (n > 0) {
+        len = (len - n) max 0
+        pos = (pos + n) % buflen
+      }
+      this
+    }
+  }
+
+  private final class DropRightIterator[A](private[this] var underlying: Iterator[A], buflen: Int) extends AbstractIterator[A] {
```

Comment:
The logic of this is pretty similar to that of `TakeRightIterator` except for the `init()`.  Can you maybe create an abstract class, and pull the conditional out so you only hit the multiply-dispatched `init` once instead of on every call?

Reply:
I don't see how. Initialization is supposed to be as lazy as possible so you have to check on every call. The use of the buffer is also quite different between both classes. There's not a lot of common code that could be factored out.

Reply:
Yeah, I guess it's not similar enough.  Never mind.